drivers/timer/cortex_m_systick.c

```diff
@@ -260,85 +309,95 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
-		delay += unannounced;
-		delay = DIV_ROUND_UP(delay, CYC_PER_TICK) * CYC_PER_TICK;
-		delay -= unannounced;
-		delay = MAX(delay, MIN_DELAY);
-		if (delay > MAX_CYCLES) {
-			last_load = MAX_CYCLES;
+		if ((int32_t)unannounced < 0) {
+			/* We haven't announced for more than half the 32-bit
+			 * wrap duration, because new timeouts keep being set
+			 * before the existing one fires.  Force an announce
+			 * to avoid loss of a wrap event, making sure the
+			 * delay is at least the minimum delay possible.
+			 */
+			last_load = MIN_DELAY;
 		} else {
-			last_load = delay;
+			const uint32_t max_cycles = max_ticks * CYC_PER_TICK;
+
+			/* Desired delay in the future */
+			delay = ticks * CYC_PER_TICK;
+
+			/* Round delay up to next tick boundary */
+			delay += unannounced;
+			delay = DIV_ROUND_UP(delay, CYC_PER_TICK) * CYC_PER_TICK;
+			delay -= unannounced;
+			delay = MAX(delay, MIN_DELAY);
+			if (delay > max_cycles) {
+				last_load = max_cycles;
+			} else {
+				last_load = delay;
+			}
 		}
-	}
 
-	val2 = SysTick->VAL;
+		val2 = sys_read32(SYST_CVR);
 
-	SysTick->LOAD = last_load - 1;
-	SysTick->VAL = 0; /* resets timer to last_load */
+		sys_write32(last_load - 1, SYST_RVR);
+		sys_write32(0, SYST_CVR); /* resets timer to last_load */
 
-	/*
-	 * Add elapsed cycles while computing the new load to cycle_count.
-	 *
-	 * Note that comparing val1 and val2 is normaly not good enough to
-	 * guess if the counter wrapped during this interval. Indeed if val1 is
-	 * close to LOAD, then there are little chances to catch val2 between
-	 * val1 and LOAD after a wrap. COUNTFLAG should be checked in addition.
-	 * But since the load computation is faster than MIN_DELAY, then we
-	 * don't need to worry about this case.
-	 */
-	if (val1 < val2) {
-		cycle_count += (val1 + (last_load_ - val2));
-	} else {
-		cycle_count += (val1 - val2);
+		/*
+		 * Add elapsed cycles while computing the new load to
+		 * cycle_count.
+		 *
+		 * Note that comparing val1 and val2 is normaly not good enough
+		 * to guess if the counter wrapped during this interval.
+		 * Indeed if val1 is close to SYST_RVR, then there are little
+		 * chances to catch val2 between val1 and SYST_RVR after a wrap.
+		 * COUNTFLAG should be checked in addition. But since the load
+		 * computation is faster than MIN_DELAY, then we don't need
+		 * to worry about this case.
+		 */
+		if (val1 < val2) {
+			cycle_count += (val1 + (last_load_ - val2));
+		} else {
+			cycle_count += (val1 - val2);
+		}
 	}
-	k_spin_unlock(&lock, key);
-#endif
 }
 
 uint32_t sys_clock_elapsed(void)
 {
-	if (!TICKLESS) {
-		return 0;
-	}
+	uint32_t ret;
 
-	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint32_t unannounced = cycle_count - announced_cycles;
-	uint32_t cyc = elapsed() + unannounced;
+	if (IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		k_spinlock_key_t key = k_spin_lock(&lock);
+		uint32_t unannounced = cycle_count - announced_cycles;
+		uint32_t cyc = elapsed() + unannounced;
```

Comment:
Please be consistent in the style used.

```suggestion
		K_SPINLOCK(&lock) {
			uint32_t unannounced = cycle_count - announced_cycles;
			uint32_t cyc = elapsed() + unannounced;
```

Reply:
I'd prefer to use K_SPINLOCK(), but it's not always possible due to the nature of the macro implementation. The compiler starts generating false positive warnings about uninitialized variables.